src/Microsoft.Extensions.Testing.Abstractions/StreamingTestExecutionSink.cs

```diff
@@ -0,0 +1,72 @@
+using System;
+using System.Collections.Concurrent;
+using System.IO;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Extensions.Testing.Abstractions
+{
+    public class StreamingTestExecutionSink : ITestExecutionSink
+    {
+        private readonly LineDelimitedJsonStream _stream;
+        private readonly ConcurrentDictionary<string, TestState> _runningTests;
+
+
+        public StreamingTestExecutionSink(Stream stream)
+        {
+            _stream = new LineDelimitedJsonStream(stream);
+            _runningTests = new ConcurrentDictionary<string, TestState>();
+        }
+        
+        public void RecordStart(Test test)
+        {
+            if (test == null)
+            {
+                throw new ArgumentNullException(nameof(test));
+            }
+
+            if (test.FullyQualifiedName != null)
+            {
+                var state = new TestState() { StartTime = DateTimeOffset.Now, };
+                _runningTests.TryAdd(test.FullyQualifiedName, state);
+            }
+
+            _stream.Send(new Message
+            {
+                MessageType = "TestExecution.TestStarted",
+                Payload = JToken.FromObject(test),
+            });
+        }
+
+        public void RecordResult(TestResult testResult)
```

Comment:
Let's name these after the message type (`TestResult`) so it's easy to match up the messages with what writes them.